File: src/codes/seanhenry/intentions/StringDecorator.java
package codes.seanhenry.intentions;

public abstract class StringDecorator {

  private StringDecorator decorator;

  public StringDecorator(StringDecorator decorator) {
    this.decorator = decorator;
  }

  public final String process(String string) {
    String decorated = decorate(string);
    if (decorator != null) {
      return decorator.process(decorated);
    }
    return decorated;
  }

  abstract public String decorate(String string);
}

File: UseCases/src/codes/seanhenry/mockgenerator/util/StringDecorator.kt
package codes.seanhenry.mockgenerator.util

abstract class StringDecorator(private val decorator: StringDecorator?) {

  fun process(string: String): String {
    val decorated = decorate(string)
    return decorator?.process(decorated) ?: decorated
  }

  protected abstract fun decorate(string: String): String
}
